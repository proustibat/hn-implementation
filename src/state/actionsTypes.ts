export const SUFFIX_FAILURE = "_FAILURE"
export const SUFFIX_SUCCESS = "_SUCCESS"
export enum storiesActionTypes {
  REQUEST_STORIES = 'REQUEST_STORIES',
  REQUEST_STORY = 'REQUEST_STORY',
}
