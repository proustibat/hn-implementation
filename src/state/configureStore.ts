import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import axios from 'axios';
import axiosMiddleware from 'redux-axios-middleware';
import thunk from 'redux-thunk';
import rootReducer from './reducers';
import { RootStateOrAny } from 'react-redux';

const client = axios.create({
  baseURL: 'https://hacker-news.firebaseio.com/v0',
  responseType: 'json',
});

const middlewares = [
  thunk,
  axiosMiddleware(client),
];

const configureStore = (preloadedState?: RootStateOrAny) => {
  return createStore(
    rootReducer,
    preloadedState,
    composeWithDevTools(applyMiddleware(...middlewares)),
  );
};

export default configureStore;
