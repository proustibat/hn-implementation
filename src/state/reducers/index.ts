import { combineReducers } from 'redux';
import stories, {StoriesReducerType} from './storiesReducer';
export type RootReducersType = {
    stories: StoriesReducerType
};

export default combineReducers({
  stories,
});
