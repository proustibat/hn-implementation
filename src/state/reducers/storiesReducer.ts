import { storiesActionTypes as Actions } from '../actionsTypes';
import { AxiosResponse } from 'axios';

export enum StoryKindType {
  job = 'job',
  story = 'story',
  comment = 'comment',
  poll = 'poll',
  pollopt = 'pollopt',
}

export type StoryProps = {
  id: number;
  deleted?: boolean; //	true if the item is deleted.
  type?: StoryKindType; // The type of item.
  by?: string; // The username of the item's author.
  time?: number; //	Creation date of the item, in Unix Time.
  text?: string; //	The comment, story or poll text. HTML.
  dead?: boolean; // true if the item is dead.
  parent?: number; // The comment's parent: either another comment or the relevant story.
  poll?: number; //	The pollopt's associated poll.
  kids?: number[]; // The ids of the item's comments, in ranked display order.
  url?: string; //	The URL of the story.
  score?: number; // The story's score, or the votes for a pollopt.
  title?: string; // The title of the story, poll or job. HTML.
  parts?: number[]; // A list of related pollopts, in display order.
  descendants?: number; //	In the case of stories or polls, the total comment count.
  loading?: boolean;
  loaded?: boolean;
};

export type StoriesReducerType = {
  items: StoryProps[];
  loading: boolean;
  loaded: boolean;
};

export const initialState: StoriesReducerType = {
  items: [],
  loading: false,
  loaded: false,
};

const LIMIT = 10;

const storiesReducer = (
  state = initialState,
  action: {
    type: Actions;
    payload?: { data?: unknown; request?: AxiosResponse };
  },
) => {
  let loadedItemId: number;
  switch (action.type) {
    case Actions.REQUEST_STORIES:
      return {
        ...state,
        loading: true,
        loaded: false,
      };
    case Actions.REQUEST_STORIES_FAILURE:
      return {
        ...state,
        loading: false,
        loaded: false,
      };
    case Actions.REQUEST_STORIES_SUCCESS:
      return {
        ...state,
        // slice temp while we don't manage queue or pagination
        items: (action?.payload?.data as number[]).slice(0, LIMIT).map(id => ({
          id,
          loaded: false,
          loading: false,
        })) as StoryProps[],
        loading: false,
        loaded: true,
      };
    case Actions.REQUEST_STORY:
      loadedItemId = action.payload?.request?.data?.id as number;
      return {
        ...state,
        items: state.items.map(item => {
          return item.id === loadedItemId
            ? {
                ...item,
                loading: true,
                loaded: false,
              }
            : item;
        }),
      };
    case Actions.REQUEST_STORY_FAILURE:
      loadedItemId = action.payload?.request?.data?.id as number;
      return {
        ...state,
        items: state.items.map(item => {
          return item.id === loadedItemId
            ? {
                ...item,
                loading: false,
                loaded: false,
              }
            : item;
        }),
      };
    case Actions.REQUEST_STORY_SUCCESS:
      const loadedItem = action.payload?.data as StoryProps;
      return {
        ...state,
        items: state.items.map(item => {
          return loadedItem && item.id === loadedItem.id
            ? {
                ...item,
                ...loadedItem,
                loading: false,
                loaded: true,
              }
            : item;
        }),
      };
    default:
      return state;
  }
};

export default storiesReducer;
