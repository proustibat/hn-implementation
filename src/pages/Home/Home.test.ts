import React from 'react';
import { render } from '@testing-library/react';
import { Home } from './index';
import { RootReducersType } from '../../state/reducers';
import { withProvider, withRouter } from '../../test-utils';
import { stories } from '../../fixtures/stories';

const stateWithStories: RootReducersType = {
  stories: {
    items: stories.map(story => ({ ...story, loaded: true, loading: false })),
    loading: false,
    loaded: true,
  },
};

const stateWithStoriesLoadingOne: RootReducersType = {
  stories: {
    items: stories.map(story => ({ ...story, loaded: story.id !== stories[1].id, loading: story.id === stories[1].id })),
    loading: false,
    loaded: false,
  },
};

const stateWithoutStories: RootReducersType = {
  stories: {
    items: [],
    loading: false,
    loaded: false,
  },
};

const stateLoadingStories: RootReducersType = {
  stories: {
    items: [],
    loading: true,
    loaded: false,
  },
};


describe('Home view', () => {
  describe.each([
    ['the store has all stories and every story are loaded', stateWithStories],
    [
      'the store has all stories and one is loading',
      stateWithStoriesLoadingOne,
    ],
    ['the store has not any stories', stateWithoutStories],
    ['the store is loading stories', stateLoadingStories],
  ])('given %s', (_, state) => {
    it('should render the view correctly by default', () => {
      // Given
      const ProvidedHome = withProvider(Home)(state);

      // When
      const { container } = render(withRouter(ProvidedHome));

      // Then
      expect(container).toMatchSnapshot();
    });
  });
});
