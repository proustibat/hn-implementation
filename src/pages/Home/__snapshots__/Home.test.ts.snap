// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home view given the store has all stories and every story are loaded should render the view correctly by default 1`] = `
<div>
  <div
    class="MuiContainer-root root_paddingTop-24 root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      home-title
    </h1>
    <div
      class="MuiBox-root MuiBox-root-1"
    >
      <div>
        <p>
          All details are loaded
        </p>
        <div
          class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14 17H4v2h10v-2zm6-8H4v2h16V9zM4 15h16v-2H4v2zM4 5v2h16V5H4z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall"
              >
                story
              </span>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-h5"
            >
              title 1
            </h2>
            <p
              class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              score
              : 
              123
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
              href="https://www.werkspot.nl"
              tabindex="0"
              target="_blank"
            >
              <span
                class="MuiButton-label"
              >
                visit
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14 17H4v2h10v-2zm6-8H4v2h16V9zM4 15h16v-2H4v2zM4 5v2h16V5H4z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall"
              >
                story
              </span>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-h5"
            >
              title 2
            </h2>
            <p
              class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              score
              : 
              456
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
              href="https://www.github.com"
              tabindex="0"
              target="_blank"
            >
              <span
                class="MuiButton-label"
              >
                visit
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14 17H4v2h10v-2zm6-8H4v2h16V9zM4 15h16v-2H4v2zM4 5v2h16V5H4z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall"
              >
                story
              </span>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-h5"
            >
              title 3
            </h2>
            <p
              class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              score
              : 
              789
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              bla bli blou
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home view given the store has all stories and one is loading should render the view correctly by default 1`] = `
<div>
  <div
    class="MuiContainer-root root_paddingTop-24 root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      home-title
    </h1>
    <div
      class="MuiBox-root MuiBox-root-2"
    >
      <div>
        <div
          class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14 17H4v2h10v-2zm6-8H4v2h16V9zM4 15h16v-2H4v2zM4 5v2h16V5H4z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall"
              >
                story
              </span>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-h5"
            >
              title 1
            </h2>
            <p
              class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              score
              : 
              123
            </p>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing"
          >
            <a
              aria-disabled="false"
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
              href="https://www.werkspot.nl"
              tabindex="0"
              target="_blank"
            >
              <span
                class="MuiButton-label"
              >
                visit
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </a>
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root rootNoLoaded_padding-0 rootNoLoaded_margin-0 MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
            role="progressbar"
          >
            <div
              class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
            />
            <div
              class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
            />
          </div>
        </div>
        <div
          class="MuiPaper-root MuiCard-root root_minWidth-275 root_paddingBottom-0.25rem root_marginBottom-0.75rem MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiChip-root root_color-#fff root_borderColor-#fff MuiChip-outlined MuiChip-sizeSmall"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiChip-icon root_color-#fff root_borderColor-#fff MuiChip-iconSmall"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M14 17H4v2h10v-2zm6-8H4v2h16V9zM4 15h16v-2H4v2zM4 5v2h16V5H4z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelSmall"
              >
                story
              </span>
            </div>
            <h2
              class="MuiTypography-root MuiTypography-h5"
            >
              title 3
            </h2>
            <p
              class="MuiTypography-root score_marginBottom-0.75rem MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              score
              : 
              789
            </p>
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              bla bli blou
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Home view given the store has not any stories should render the view correctly by default 1`] = `
<div>
  <div
    class="MuiContainer-root root_paddingTop-24 root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      home-title
    </h1>
    <div
      class="MuiBox-root MuiBox-root-3"
    >
      <div>
        no-content
      </div>
    </div>
  </div>
</div>
`;

exports[`Home view given the store is loading stories should render the view correctly by default 1`] = `
<div>
  <div
    class="MuiContainer-root root_paddingTop-24 root_paddingBottom-96 MuiContainer-maxWidthSm"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      home-title
    </h1>
    <div
      class="MuiBox-root MuiBox-root-4"
    >
      <div
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
        role="progressbar"
      >
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
        />
        <div
          class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
        />
      </div>
    </div>
  </div>
</div>
`;
